src/dials/algorithms/correlation/analysis.py

```diff
@@ -0,0 +1,211 @@
+            for d in individual_merged_intensities
+        ]
+
+        self.params.lattice_group = self.datasets[0].space_group_info()
+        self.params.space_group = self.datasets[0].space_group_info()
+
+        self.cosym_analysis = CosymAnalysis(self.datasets, self.params)
+
+    def _filter_min_reflections(self, experiments, reflections):
+        """
+        Filter all datasets that have less than the specified number of reflections.
+        """
+        identifiers = []
+
+        for expt, refl in zip(experiments, reflections):
+            if len(refl) >= self.params.min_reflections:
+                identifiers.append(expt.identifier)
+
+        return select_datasets_on_identifiers(
+            experiments, reflections, use_datasets=identifiers
+        )
+
+    def calculate_matrices(self):
+        """
+        Calculates the cc and cos angle correlation matrices for the input dataset.
+        The cc matrix is converted to a distance matrix, and both are condensed into a nC2 array.
+        Matrices are converted to plotly json files for visual output.
+        """
+
+        self.cosym_analysis._intialise_target()
+        self.cosym_analysis._determine_dimensions()
+        self.cosym_analysis._optimise(
+            self.cosym_analysis.params.minimization.engine,
+            max_iterations=self.cosym_analysis.params.minimization.max_iterations,
+            max_calls=self.cosym_analysis.params.minimization.max_calls,
+        )
+
+        logger.info("\nCalculating Correlation Matrix (rij matrix - see dials.cosym)\n")
+
+        correlation_matrix = self.cosym_analysis.target.rij_matrix
+
+        for i in range(correlation_matrix.shape[0]):
+            correlation_matrix[i, i] = 1
+
+        # clip values of correlation matrix to account for floating point errors
+        correlation_matrix[np.where(correlation_matrix < -1)] = -1
+        correlation_matrix[np.where(correlation_matrix > 1)] = 1
+        diffraction_dissimilarity = 1 - correlation_matrix
+
+        assert ssd.is_valid_dm(diffraction_dissimilarity, tol=1e-12)
+        # convert the redundant n*n square matrix form into a condensed nC2 array
+        cc_dist_mat = ssd.squareform(diffraction_dissimilarity, checks=False)
+
+        cc_linkage_matrix = hierarchy.linkage(cc_dist_mat, method="average")
+
+        logger.info(
+            "Calculating Cos Angle Matrix from optimised cosym coordinates (see dials.cosym)\n"
+        )
+
+        cos_dist_mat = ssd.pdist(self.cosym_analysis.coords, metric="cosine")
+        cos_angle = 1 - ssd.squareform(cos_dist_mat)
+        cos_linkage_matrix = hierarchy.linkage(cos_dist_mat, method="average")
+
+        self.cc_linkage_matrix = cc_linkage_matrix
+        self.correlation_matrix = correlation_matrix
+        self.cos_angle = cos_angle
+        self.cos_linkage_matrix = cos_linkage_matrix
+
+        labels = list(range(0, len(self._experiments)))
+
+        self.cc_json = to_plotly_json(
```

Comment:
Just thinking about the logical separtion in tandem with the command line program - i think there are three parts - 1. set up things needed for the calculations (in the init ✅ ), 2. run the cosym analysis and matrix operations + linkages, 3. generate output for saving to json/html etc. With the idea that if you wanted to just run the analysis without calculating things for output then this is separate.

Reply:
Great suggestion, have addressed in f484d11 to make the program easy to run without needing any json/html output.